File: aite/contracts/lock.py
"""契约锁：记录并校验 aite/contracts/ 下每个契约文件的 sha256。

    python -m aite.contracts.lock --write    # 生成 / 刷新仓库根的 .contracts.lock
    python -m aite.contracts.lock --check    # 一致 -> "OK <n> files" 退出 0；否则打印差异退出 1

本文件自身与 __pycache__ 不参与锁定：lock.py 是工具不是契约，改它不构成契约变更。
"""
import argparse
import hashlib
import sys
from pathlib import Path

CONTRACTS_DIR = Path(__file__).resolve().parent
REPO_ROOT = CONTRACTS_DIR.parents[1]
LOCK_PATH = REPO_ROOT / ".contracts.lock"
SELF_NAME = Path(__file__).resolve().name

HEADER = "# aite contracts lock — sha256 of aite/contracts/*.py（lock.py 除外）"


def contract_files() -> list[Path]:
    """参与锁定的契约文件，按仓库相对路径排序。

    收**所有**文件而不只是 *.py：派单原文是「每个 contracts 文件的 sha256」。
    只锁 .py 的话，往 aite/contracts/ 丢一个 .pyi 覆盖类型、或塞个 .json 进去，
    --check 照样报 OK，保护面就有洞。排除项只有 lock.py 自身与 __pycache__。
    """
    out = []
    for p in CONTRACTS_DIR.rglob("*"):
        if not p.is_file():
            continue
        if p.name == SELF_NAME and p.parent == CONTRACTS_DIR:
            continue
        if "__pycache__" in p.parts:
            continue
        out.append(p)
    return sorted(out, key=lambda p: p.relative_to(REPO_ROOT).as_posix())


def digest(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()


def current() -> dict[str, str]:
    return {p.relative_to(REPO_ROOT).as_posix(): digest(p) for p in contract_files()}


def render(entries: dict[str, str]) -> str:
    body = "".join(f"{h}  {rel}\n" for rel, h in sorted(entries.items()))
    return HEADER + "\n" + body


def parse(text: str) -> dict[str, str]:
    entries = {}
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        h, _, rel = line.partition("  ")
        if not rel:
            raise ValueError(f"无法解析的锁行：{line!r}")
        entries[rel.strip()] = h.strip()
    return entries


MIN_CONTRACT_FILES = 10   # §3.1/§3.2 的十个模块，加 __init__.py 共 11；少于这个数一定是出事了


def _guard_not_empty(entries: dict[str, str]) -> None:
    """空表守卫：没有这条，contract_files() 返空时 --check 会打印 `OK 0 files` 退出 0。

    「锁校验绿」是 §2.3 C2 给所有 track 的回归闸门，绿得毫无内容比红了更危险。"""
    if len(entries) < MIN_CONTRACT_FILES:
        raise SystemExit(
            f"契约文件只找到 {len(entries)} 个（期望 >= {MIN_CONTRACT_FILES}）："
            f"{sorted(entries)}\n{CONTRACTS_DIR} 是不是被删了或路径解析错了？"
        )


def cmd_write() -> int:
    entries = current()
    _guard_not_empty(entries)
    LOCK_PATH.write_text(render(entries), encoding="utf-8")
    print(f"wrote {len(entries)} files -> {LOCK_PATH.relative_to(REPO_ROOT).as_posix()}")
    return 0


def cmd_check() -> int:
    entries = current()
    _guard_not_empty(entries)
    if not LOCK_PATH.exists():
        print(f"MISSING {LOCK_PATH.relative_to(REPO_ROOT).as_posix()} 不存在，先跑 --write", file=sys.stderr)
        return 1
    locked = parse(LOCK_PATH.read_text(encoding="utf-8"))
    diffs = []
    for rel in sorted(set(locked) | set(entries)):
        if rel not in entries:
            diffs.append(f"  deleted  {rel}（在锁里但文件不存在）")
        elif rel not in locked:
            diffs.append(f"  added    {rel}（新文件未入锁）")
        elif locked[rel] != entries[rel]:
            diffs.append(f"  changed  {rel}\n    locked {locked[rel]}\n    actual {entries[rel]}")
    if diffs:
        print(f"MISMATCH {len(diffs)} file(s):", file=sys.stderr)
        for d in diffs:
            print(d, file=sys.stderr)
        return 1
    print(f"OK {len(entries)} files")
    return 0


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(prog="python -m aite.contracts.lock", description=__doc__)
    g = ap.add_mutually_exclusive_group(required=True)
    g.add_argument("--write", action="store_true", help="生成 / 刷新 .contracts.lock")
    g.add_argument("--check", action="store_true", help="校验 .contracts.lock 与实际文件一致")
    args = ap.parse_args(argv)
    return cmd_write() if args.write else cmd_check()


if __name__ == "__main__":
    sys.exit(main())

File: core/crates/app/src/lock.rs
//! 契约锁：记录并校验契约文件的 sha256（对应旧 aite/contracts/lock.py）。
//!
//!   aite contracts lock --write    # 生成 / 刷新仓库根的 .contracts.lock
//!   aite contracts lock --check    # 一致 -> "OK <n> files" 退出 0；否则打印差异退出 1
//!
//! 锁定面：proto/aite/v1/**、core/crates/contracts/**（Cargo.toml + src + tests）、aite/contracts/**（Python，RΩ 删除时一并重锁）。
//! 收目录下**所有**文件而不只是源码：往契约目录丢任何东西都会让 --check 变红，保护面没有洞。
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result, bail};
use sha2::{Digest, Sha256};

pub const LOCK_FILE: &str = ".contracts.lock";
pub const LOCK_DIRS: &[&str] = &["proto/aite/v1", "core/crates/contracts", "aite/contracts"];
const HEADER: &str = "# aite contracts lock — sha256 of proto/aite/v1/**, core/crates/contracts/**, aite/contracts/**（lock.py/__pycache__/target 除外）";
/// 五份 proto + contracts crate（Cargo.toml + 13 源 + 7 测试）+ Python 11 份；少于这个数一定是出事了
pub const MIN_CONTRACT_FILES: usize = 20;

fn skip(rel: &Path) -> bool {
    rel.components().any(|c| {
        let s = c.as_os_str().to_string_lossy();
        s == "__pycache__" || s == "target" || s == ".DS_Store"
    }) || rel.ends_with("aite/contracts/lock.py")
}

pub fn find_repo_root(explicit: Option<PathBuf>) -> Result<PathBuf> {
    if let Some(p) = explicit {
        return p
            .canonicalize()
            .with_context(|| format!("--repo {} 不存在", p.display()));
    }
    let mut dir = std::env::current_dir()?;
    loop {
        if dir.join("proto/aite/v1").is_dir() && dir.join("core/Cargo.toml").is_file() {
            return Ok(dir);
        }
        if !dir.pop() {
            bail!(
                "从当前目录向上找不到仓库根（要有 proto/aite/v1 与 core/Cargo.toml）；可用 --repo 指定"
            );
        }
    }
}

fn walk(dir: &Path, out: &mut Vec<PathBuf>) -> Result<()> {
    for entry in fs::read_dir(dir).with_context(|| format!("读目录 {}", dir.display()))? {
        let p = entry?.path();
        if p.is_dir() {
            walk(&p, out)?;
        } else if p.is_file() {
            out.push(p);
        }
    }
    Ok(())
}

pub fn current(root: &Path) -> Result<BTreeMap<String, String>> {
    let mut files = Vec::new();
    for d in LOCK_DIRS {
        let dir = root.join(d);
        if dir.is_dir() {
            walk(&dir, &mut files)?;
        }
    }
    let mut out = BTreeMap::new();
    for p in files {
        let rel = p.strip_prefix(root)?.to_path_buf();
        if skip(&rel) {
            continue;
        }
        let bytes = fs::read(&p)?;
        out.insert(
            rel.to_string_lossy().replace('\\', "/"),
            hex::encode(Sha256::digest(&bytes)),
        );
    }
    Ok(out)
}

pub fn render(entries: &BTreeMap<String, String>) -> String {
    let mut s = String::from(HEADER);
    s.push('\n');
    for (rel, h) in entries {
        s.push_str(&format!("{h}  {rel}\n"));
    }
    s
}

pub fn parse(text: &str) -> Result<BTreeMap<String, String>> {
    let mut out = BTreeMap::new();
    for line in text.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let (h, rel) = line
            .split_once("  ")
            .with_context(|| format!("无法解析的锁行：{line:?}"))?;
        out.insert(rel.trim().to_string(), h.trim().to_string());
    }
    Ok(out)
}

fn guard_not_empty(entries: &BTreeMap<String, String>) -> Result<()> {
    if entries.len() < MIN_CONTRACT_FILES {
        bail!(
            "契约文件只找到 {} 个（期望 >= {MIN_CONTRACT_FILES}）：{:?}\n锁定目录是不是被删了或仓库根找错了？",
            entries.len(),
            entries.keys().collect::<Vec<_>>()
        );
    }
    Ok(())
}

pub fn run(write: bool, repo: Option<PathBuf>) -> Result<i32> {
    let root = find_repo_root(repo)?;
    let entries = current(&root)?;
    guard_not_empty(&entries)?;
    let lock_path = root.join(LOCK_FILE);
    if write {
        if std::env::var("AITE_RELOCK").as_deref() != Ok("1") {
            bail!("--write 需要 AITE_RELOCK=1（只有总管 / R0 该重锁契约）");
        }
        fs::write(&lock_path, render(&entries))?;
        println!("wrote {} files -> {LOCK_FILE}", entries.len());
        return Ok(0);
    }
    if !lock_path.exists() {
        eprintln!("MISSING {LOCK_FILE} 不存在，先跑 --write");
        return Ok(1);
    }
    let locked = parse(&fs::read_to_string(&lock_path)?)?;
    let mut diffs = Vec::new();
    let keys: std::collections::BTreeSet<&String> = locked.keys().chain(entries.keys()).collect();
    for rel in keys {
        match (locked.get(rel), entries.get(rel)) {
            (Some(_), None) => diffs.push(format!("  deleted  {rel}（在锁里但文件不存在）")),
            (None, Some(_)) => diffs.push(format!("  added    {rel}（新文件未入锁）")),
            (Some(l), Some(a)) if l != a => {
                diffs.push(format!("  changed  {rel}\n    locked {l}\n    actual {a}"))
            }
            _ => {}
        }
    }
    if !diffs.is_empty() {
        eprintln!("MISMATCH {} file(s):", diffs.len());
        for d in diffs {
            eprintln!("{d}");
        }
        return Ok(1);
    }
    println!("OK {} files", entries.len());
    Ok(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_render_round_trip() {
        let mut m = BTreeMap::new();
        m.insert("proto/aite/v1/a.proto".to_string(), "ab".repeat(32));
        let text = render(&m);
        assert!(text.starts_with("# aite contracts lock"));
        assert_eq!(parse(&text).unwrap(), m);
    }

    #[test]
    fn skips_tooling_and_caches() {
        assert!(skip(Path::new("aite/contracts/lock.py")));
        assert!(skip(Path::new("aite/contracts/__pycache__/x.pyc")));
        assert!(!skip(Path::new("core/crates/contracts/src/lib.rs")));
    }
}
